### PR TITLE
qemu_guest_agent: Check new point 'file system usage statistics'

### PR DESCRIPTION
**Pull request by 6-dehan**
Add checkpoint 'file system usage statistics' and check it.

ID: 1857505
Signed-off-by: Dehan Meng  \<demeng@redhat.com\>

### CODE REVIEW

File: qemu/tests/qemu_guest_agent.py [hunk 1]
```diff
@@ -2550,6 +2551,11 @@ def gagent_check_fsinfo(self, test, params, env):
             if params["os_type"] == "windows":
                 mount_pt = mount_pt[:2]
 
+            error_context.context("Check filesystem usage statistics .", logging.info)
+            disk_total_bytes = fs["total-bytes"]
+            disk_used_bytes = fs["used-bytes"]
+            logging.info("The total_bytes of %s is %s,The used_bytes of %s is %s." % (mount_pt, disk_total_bytes, mount_pt, disk_used_bytes))
+
```
Comment (xiagao) [inline qemu/tests/qemu_guest_agent.py:2558]: Check point is missing, you should check if the value is the same with it in guest.

Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:2558]: oh,right,re-editing.

**Comment by 6-dehan** [comment]
(2/2) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_fsinfo.q35: PASS (161.19 s)

**Comment by 6-dehan** [comment]
@xiagao

**Comment by xiagao** [comment]
For a simple patch, you don't need so many commits.
One is enough.

**Comment by 6-dehan** [comment]
> For a simple patch, you don't need so many commits.
> One is enough.

okay,I see.thanks

**Review by xiagao — CHANGES_REQUESTED** [review]

File: qemu/tests/qemu_guest_agent.py [hunk 2]
```diff
@@ -2550,6 +2551,52 @@ def gagent_check_fsinfo(self, test, params, env):
             if params["os_type"] == "windows":
                 mount_pt = mount_pt[:2]
 
+            def get_guest_disk_usage(mountpoint):
```
[the remark this answers is not in the document]
Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:2554]: you mean set these two sub function before gagent_check_fsinfo()?

File: qemu/tests/qemu_guest_agent.py [hunk 3]
```diff
@@ -2550,6 +2551,52 @@ def gagent_check_fsinfo(self, test, params, env):
...
+                :param mount_point: the mountpoint of disk device as the condition
+                to get corresponding value
+                '''
+
+                disk_usage_guest = get_guest_disk_usage(mount_point)
+                disk_total_qga = int(fs["total-bytes"])
+                disk_used_qga = int(fs["used-bytes"])
+                if disk_total_qga != disk_usage_guest[0] or disk_used_qga != disk_usage_guest[1]:
+                    test.fail("File System Total bytes or Used bytes doesn't match.\n"
+                              "total_bytes from guest-agent is %d.\ntotal_bytes from guest os is %d.\n"
```
[the remark this answers is not in the document]
Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:2580]: okay,I'll edit it asap.

File: qemu/tests/qemu_guest_agent.py [hunk 4]
```diff
@@ -2550,6 +2551,52 @@ def gagent_check_fsinfo(self, test, params, env):
...
+                                 % (mount_pt, disk_total_qga, mount_pt, disk_used_qga))
+
+            error_context.context("Check filesystem usage statistics .", logging.info)
+            if mount_pt != 'C:':
+                if mount_pt != 'Sy':
+                    check_diskusage_guest_qga(mount_pt)
+                else:
+                    s_mount_pt = 'S:'
+                    check_diskusage_guest_qga(s_mount_pt)
+            else:
```
[the remark this answers is not in the document]
Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:2594]: Worth to say,the C:\\ system disk's capacity is constantly changing,so We cannot get a same number from guest and qga,so,it no need to compare between qga and guest.

File: qemu/tests/qemu_guest_agent.py [hunk 5]
```diff
@@ -2550,6 +2551,52 @@ def gagent_check_fsinfo(self, test, params, env):
...
+                              "total_bytes from guest-agent is %d.\ntotal_bytes from guest os is %d.\n"
+                              "used_bytes from guest is %d.\nused_bytes from guest os is %d."
+                              % (disk_total_qga, disk_usage_guest[0], disk_used_qga, disk_usage_guest[1]))
+                else:
+                    logging.info("The total_bytes of %s is %d, The used_bytes of %s is %d."
+                                 % (mount_pt, disk_total_qga, mount_pt, disk_used_qga))
+
+            error_context.context("Check filesystem usage statistics .", logging.info)
+            if mount_pt != 'C:':
+                if mount_pt != 'Sy':
```
[the remark this answers is not in the document]
Comment (6-dehan) [reply qemu/tests/qemu_guest_agent.py:2589]: 'Sy' is not a reasonable string for getting the real usage from guest,there are no 'Sy' Keyword from guest_cmd.

File: qemu/tests/qemu_guest_agent.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py:2554]: > you mean set these two sub function before gagent_check_fsinfo()?

no ,still in this function,but put it on the beginning.

File: qemu/tests/qemu_guest_agent.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py:2589]: So if this string is not reasonable, we should not put it on automation.

File: qemu/tests/qemu_guest_agent.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py:2594]: You mean the value is always changing? which is weird.
And we should get the same value from qga and guest. If not, if it's a product bug?

**Comment by 6-dehan** [comment]
This string "Sy" is intercepted by command before I edited it .we just can
set if-else statement to classify it .

**Comment by xiagao** [comment]
> This string "Sy" is intercepted by command before I edited it .we just can set if-else statement to classify it .
> […](#)
Then we can handle it in the before code.

**Comment by 6-dehan** [comment]
(1)There are some problems with the actual situations that I cannot find 'S:' by command `wmic volume get /vaule` in guest(win2019_ovmf or win2016_ovmf) .and there are 'S:' and 'System Reserved' two keypoint from test result log.
(2)The usage of disk 'C:' should be considered specially,Need to run lots of times to find out the interregional value.

File: qemu/tests/qemu_guest_agent.py [hunk 6]
```diff
@@ -2543,13 +2544,88 @@ def gagent_check_fsinfo(self, test, params, env):
...
         error_context.context("Check all file system info in a loop.", logging.info)
         fs_info_qga = self.gagent.get_fsinfo()
         for fs in fs_info_qga:
             mount_pt = fs["mountpoint"]
-            if params["os_type"] == "windows":
+            if (params["os_type"] == "windows" and
+                    mount_pt != "System Reserved"):
                 mount_pt = mount_pt[:2]
 
+            error_context.context("Check filesystem usage statistics .",
```
Comment (xiagao) [inline qemu/tests/qemu_guest_agent.py:2594]: error_context.context("Check filesystem usage statistics between qga and guest.",

File: qemu/tests/qemu_guest_agent.py [hunk 7]
```diff
@@ -2543,13 +2544,88 @@ def gagent_check_fsinfo(self, test, params, env):
...
+                                  logging.info)
+            if mount_pt != 'C:' and mount_pt != '/':
+                if mount_pt != 'System Reserved' and mount_pt != 'S:':
+                    check_diskusage_qga_guest(mount_pt)
+                else:
+                    # disk usage statistics for System Reserved
+                    # volumes is not supported.
```
Comment (xiagao) [inline qemu/tests/qemu_guest_agent.py:2601]: You mean mount_pt == 'System Reserved' or 'S:',
both of the two disks didn't support disk usage test from product?

File: qemu/tests/qemu_guest_agent.py [hunk 8]
```diff
@@ -2543,13 +2544,88 @@ def gagent_check_fsinfo(self, test, params, env):
...
+                    pass
+            else:
+                # file system 'C:' and '/' need to
+                # count with special method.
+                disk_usage_guest = get_guest_disk_usage(mount_pt)
+                disk_total_guest = int(disk_usage_guest[0])
```
Comment (xiagao) [inline qemu/tests/qemu_guest_agent.py:2607]: could you compelete int() in get_guest_disk_usage() function?

File: qemu/tests/qemu_guest_agent.py [hunk 9]
```diff
@@ -2543,13 +2544,88 @@ def gagent_check_fsinfo(self, test, params, env):
...
+            if (disk_total_qga != disk_usage_guest[0] or
+                    disk_used_qga != disk_usage_guest[1]):
+                test.fail("File System Total bytes or Used bytes doesn't match.\n"
+                          "total_bytes from qga_cmd is %d."
+                          "total_bytes from guest is %d.\n"
+                          "used_bytes from qga_cmd is %d."
+                          "used_bytes from guest is %d."
+                          % (disk_total_qga, disk_usage_guest[0],
+                             disk_used_qga, disk_usage_guest[1]))
+            else:
```
Comment (xiagao) [inline qemu/tests/qemu_guest_agent.py:2581]: There is no need to add this else as we can check them in other logs and it's not the checkpoint.

File: qemu/tests/qemu_guest_agent.py [hunk 10]
```diff
@@ -2543,13 +2544,88 @@ def gagent_check_fsinfo(self, test, params, env):
...
             mount_pt = fs["mountpoint"]
-            if params["os_type"] == "windows":
+            if (params["os_type"] == "windows" and
+                    mount_pt != "System Reserved"):
                 mount_pt = mount_pt[:2]
 
+            error_context.context("Check filesystem usage statistics"
+                                  "between guest and qga.", logging.info)
+            if mount_pt != 'C:' and mount_pt != '/':
+                if mount_pt != 'System Reserved' and mount_pt != 'S:':
```
Comment (xiagao) [inline qemu/tests/qemu_guest_agent.py:2597]: About this part, we can wait until the product part are ready.

File: qemu/tests/qemu_guest_agent.py [hunk 11]
```diff
@@ -2543,13 +2544,88 @@ def gagent_check_fsinfo(self, test, params, env):
...
+                    # Disk C: usage have a floating interval,
+                    # so we set it as '10485760'.
+                    test.fail("File System Total bytes or Used bytes doesn't match.\n"
+                              "total_bytes from qga_cmd is %d."
+                              "total_bytes from guest is %d.\n"
+                              "used_bytes from qga_cmd is %d."
+                              "used_bytes from guest is %d."
+                              % (disk_total_qga, disk_usage_guest[0],
+                                 disk_used_qga, disk_usage_guest[1]))
+                else:
```
Comment (xiagao) [inline qemu/tests/qemu_guest_agent.py:2624]: There is no need to add else as above.

**Comment by 6-dehan** [comment]
Need to create new repo for these patches.

**Pull request closed without merging**